### DISCUSSION TITLE
ESP32-CAM unable to use machine.UART with wifi and camera (micropython)

### DISCUSSION THREAD

**Discussion by boylucky** [micropython/micropython#14320] [ESP32] [answered]
I am having troubles with ESP32-CAM module running camera and wifi with machine.UART on any GPIO (tested with GPIO1 rx, GPIO3 tx and GPIO14 rx, GPIO16 tx). I would need to communicate via UART with mini MP3 module MH2024K-24SS. I am getting ERROR:

Guru Meditation Error: Core  1 panic'ed (StoreProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x40093470  PS      : 0x00060033  A0      : 0x80091054  A1      : 0x3ffcba60  
A2      : 0x0000004b  A3      : 0xb33fffff  A4      : 0x0000cdcd  A5      : 0x0000abab  
A6      : 0x00060023  A7      : 0x00060020  A8      : 0x0000abab  A9      : 0xffffffff  
A10     : 0x3ffcbcc8  A11     : 0x00060a23  A12     : 0x00060a20  A13     : 0x0000abab  
A14     : 0x003fffff  A15     : 0x00060620  SAR     : 0x00000020  EXCCAUSE: 0x0000001d  
EXCVADDR: 0x0000004b  LBEG    : 0x4008dafa  LEND    : 0x4008db05  LCOUNT  : 0x00000000  

Backtrace: 0x4009346d:0x3ffcba60 0x40091051:0x3ffcbaa0 0x4012327c:0x3ffcbae0 0x400d6df9:0x3ffcbb10 0x400d6f30:0x3ffcbb30 0x400ed59e:0x3ffcbb50 0x400ed6ea:0x3ffcbbe0 0x400d6caa:0x3ffcbc30

This error appear imediately when executing following part of code:
def __init__(self, pin_TX, pin_RX): #pin_RX
    self.uart = UART(1, 9600, tx=pin_TX, rx=pin_RX)  - with above GPIOs

I am getting a bit crazy from that :o) I tried to turn off the debug messages also - esp.osdebug(None) to turn off the the REPL messages, also without success. The module communicate on baud rate 9600. I was also trying to disable the camera and wifi befor calling the UART init, also with no luck.

Just to explain a bit my goal. I am building project where I need to use camera, wifi and audio. Because I am not able to run MP3 or wav files directly on ESP32-CAM I wanted to use mini MP3 module with SD card and all needed audio stored there. Then I just need to call the MP3 to play specific MP3 file which is possible to do via the UART. But as mentioned I always get the ERROR above when initiating the UART. When I run the code separately (only the code which do not use camera and wifi, it only send needed commands via UART) then it work without problems. But when I want to use it together I am getting problems. In case there is no real ability to do what I want with this approach, would be there a possibility to somehow run it without UART? Like directly on some GPIO? I was trying that like simulate the UART but I think it is impossible and this is probably stupid idea. Any help or your experiences will be appreciated. I am thinking one more approach now as the module MH2024K-24SS have also pins for next and previous mp3 file to play. So will give it a try to connect with GPIO pin and send required number of nexts :o) . Probably also a bit crazy but I will give it a try. Thanks if anybody would have better idea or any experiences which could solve the problem with UART. Thanks...

**Comment by boylucky** [comment]
I have tested the idea mentioned in previous post to use the pin on MH2024K-24SS to play next song (ADKEY1 with 15Kohm resistor) but there is needed quite long push to let the module triger the push. So I am not able to use it for more then 100 mp3 files to directly play specific song by correct number of push via GPIO PIN on ESP32-CAM. Next possibility would be probably to add something like ATTINY to drive the communication with the MH2024K-24SS based on inputs from GPIO PIN on ESP32-CAM. But it is getting stranger and more nonsense I would say.

Some time ago I testet project from this web site: [link](https://www.hackster.io/RoboticaDIY/esp8266-how-to-make-wi-fi-radio-5bee14) . Maybe it could be the right approach to try something similar but with use of micropython as I have the rest of the code already writen in micropython. I could use GPIO 16 on ESP32-CAM which is the RX pin and is not used by any device (wifi, camera, SD card) and the mp3 files could be stored and red from the SD card placed in ESP32-CAM. Would anybody have some idea about the micropython code which could work?

Any comments or ideas are appreciated. Thanks.

**Reply by davefes** [reply]
Which ESP32-CAM board are you using, most of those pins are not visible on my board?

**Reply by davefes** [reply]
And which image are you using?

**Reply by boylucky** [reply]
I am using ESP32-S board (I would say ESP32-CAM of AI-Thinker clone) bought here: [link](https://www.aliexpress.com/item/1005006299363624.html?spm=a2g0o.order_detail.order_detail_item.3.201ef19ckxNRK6)

And I am using following image of micropython: [link](https://github.com/shariltumin/esp32-cam-micropython-2022/tree/main/X23/esp32-aiThinker/firmwares/wifi%2Bssl)

**Comment by shariltumin** [comment, marked as answer]
Please try PIN12 and PIN13. Below is the REPL log to show that the setup is working for UART1, SDcard and camera, all active together:

```
MicroPython v1.20.0-206-g33b403dfb-kaki5 on 2024-01-19; ESP32 AIT OV2640 (KAKI5) with ESP32
>>> import machine
>>> u1 = machine.UART(1, tx=12, rx=13)
>>> u1.
any             read            readinto        readline
write           CTS             INV_CTS         INV_RTS
INV_RX          INV_TX          RTS             deinit
flush           init            sendbreak       txdone
>>> u1.init()
>>> u1.write(b'fdsfsdfsdfsdfsdfdsffdssdf')
25
>>> u1.read()
>>> 
>>> u1
UART(1, baudrate=115201, bits=8, parity=None, stop=1, tx=12, rx=13, rts=-1, cts=-1, txbuf=256, rxbuf=256, timeout=0, timeout_char=0)
>>> import camera
>>> camera.init()
True
>>> sd = machine.SDCard()
>>> os.mount(sd, "/sd")
>>> os.listdir("/sd")
['pic-100.jpg', 'pic-000.jpg', 'pic-001.jpg', 'pic-002.jpg
... 'pic099.jpg', 'test']
>>> img = camera.capture()
>>> len(img)
10920
>>>
```

Give it a try and hope it works for your project.

**Reply by boylucky** [reply]
Thanks a lot for your help. It works! Looks that there is only problem to don't have connected the tx GPIO 12 and rx GPIO13 during the boot of ESP32 othervise it do not start. So I need to solve it somehow (actually it is enough to have connected the TX pin) so I will do more tests if any solution which would work with the PIN connected all the time.

I spend 2 days of many tries with the UART and you solve it with this easy solution. Really thanks.  I did not try that before because I red somewhere about ESP32-CAM to avoid use of those pins when camera is used. Thank you also for your work you did on the firmware for ESP32-CAM. Really great work and I appreciate it.

**Comment by davefes** [comment]
Are those pins that you are referring to on the downloader board?

> So I need to solve it somehow (actually it is enough to have connected the TX pin) so I will do more tests if any solution which would work with the PIN connected all the time.

Enough to cause the problem?  On the ESP32 Generic board GPIO12 is a strapping pin and might require special treatment.

**Reply by boylucky** [reply]
For now it looks it cause problem only in case I push the reboot button on the board to reset the ESP32-CAM. So will do more tests tomorrow to see if it will be ok when the module is powered off and on completely. Then it would be ok for me to keep it without modification. Will post the result here.

**Reply by davefes** [reply]
Checked the ESP32-CAM schematic and pins on that header:

My tests on the ESP32 Generic
GPIO26 probably OK
GPIO2 is a strapping pin
GPIO14 I see it go hi during deepsleep
GPIO15 I see it go hi during deepsleep and it is a strapping pin
GPIO13 probably OK
GPIO12 is a strapping pin

**Reply by boylucky** [reply]
Thanks for your tests. GPIO13 looks great for TX with my tests currently done. I need to do further development to see how it works further. I will keep posted here.
